core/src/main/java/com/linecorp/armeria/common/Flags.java

```diff
@@ -1422,87 +1306,72 @@ private static ExceptionVerbosity exceptionLoggingMode(String name, String defau
-            } catch (Exception e) {
-                // null or non-integer
-                return false;
-            }
-        }));
-    }
-
-    private static long getLong(String name, long defaultValue, LongPredicate validator) {
-        return Long.parseLong(getNormalized(name, StringUtil.toString(defaultValue), value -> {
-            try {
-                return validator.test(Long.parseLong(value));
-            } catch (Exception e) {
-                // null or non-integer
-                return false;
-            }
-        }));
-    }
-
-    private static String get(String name, String defaultValue, Predicate<String> validator) {
+    private static String getNormalized(String name, String defaultValue,
+                                        Predicate<String> validator) {
         final String fullName = PREFIX + name;
-        final String value = System.getProperty(fullName);
-        if (value == null) {
-            logger.info("{}: {} (default)", fullName, defaultValue);
-            return defaultValue;
+        String value = System.getProperty(fullName);
+        if (value != null) {
+            value = Ascii.toLowerCase(value);
         }
 
-        if (validator.test(value)) {
-            logger.info("{}: {}", fullName, value);
-            return value;
+        if (value != null) {
+            if (validator.test(value)) {
+                logger.info("{}: {} (JVM option)", fullName, value);
+                return value;
+            }
+            logger.warn("{}: {} (JVM option) fail validation", fullName, value);
         }
-
-        logger.info("{}: {} (default instead of: {})", fullName, defaultValue, value);
+        logger.info("{}: {} (default)", fullName, defaultValue);
         return defaultValue;
     }
 
-    private static String getNormalized(String name, String defaultValue, Predicate<String> validator) {
-        final String fullName = PREFIX + name;
-        final String value = getLowerCased(fullName);
-        if (value == null) {
-            logger.info("{}: {} (default)", fullName, defaultValue);
-            return defaultValue;
-        }
+    private static <T> T getValue(Function<FlagsProvider, @Nullable T> method, String flagName) {
+        return getValue(method, flagName, unused -> true);
+    }
 
-        if (validator.test(value)) {
-            logger.info("{}: {}", fullName, value);
-            return value;
-        }
+    private static <T> T getValue(Function<FlagsProvider, @Nullable T> method,
+                                  String flagName, Predicate<T> validator) {
+        for (FlagsProvider provider : FLAGS_PROVIDERS) {
+            try {
+                final T value = method.apply(provider);
+                if (provider instanceof DefaultFlagsProvider) {
+                    logger.info("{}: {} ({})", flagName, value, provider.getClass().getSimpleName());
+                    return value;
+                }
```

Comment:
Is this `if` to skip validation at runtime? If so, how about removing this block so they are validated just in case we make any mistakes?

Reply:
> Is this if to skip validation at runtime? 

Yes, and also to skip validation for null default value `preferredIpV4Addresses`.
After check the useage of `Flags.preferredIpV4Addresses()`, I think I can change the default `preferredIpV4Addresses` from null to predicate that always return true. So that, I can remove default value validation skipping as you suggested.

Reply:
Sounds good to me!